### ISSUE TITLE
Computation discrepancy between old and new GPU backends

### ISSUE THREAD

**Issue by ozancaglayan** [Theano/Theano#5475]
I've tested the current master and the libgpuarray with my code on 2 different GPUs and for both backends.

 - On a Quadro K5200 (Kepler, SM=3.5), both backends reproduced equivalent results.
 - On a GTX980 (Maxwell, SM=5.2), gpuarray backend produces completely nonsense batch losses and the training diverges while old backend works without problem.

How can we debug this?

**Comment by nouiz** [comment]
Can you confirm you use the dev version of Theano and libgpuarray? How
"old" they are?

There is a PR that is a fix: https://github.com/Theano/Theano/pull/5469

Can you try it? Maybe it will fix you r problem.

Otherwise, you will need to run with the flag mode=DebugMode.

**Comment by ozancaglayan** [comment]
I've pulled Theano now and applied the PR. libgpuarray was already at the latest. I'm still having the same behaviour.

I'll now check with disabling optimizations and DebugMode as well.

**Comment by nouiz** [comment]
To help you find faster, first try with optimizer=fast_compile.

If that fix it, then doing `mode=DebugMode,DebugMode.check_py=False` would
be the next step, it is much slower.

**Comment by ozancaglayan** [comment]
optimizer=fast_compile does not fix but optimizer=None is **OK**. Will now try with DebugMode.

**Comment by HenryJia** [comment]
Also getting computational discrepancies running Theano with new backend with Keras of factors of 2 times or greater.

I'm using GTX 1080 (Pascal architecture)

**Comment by ozancaglayan** [comment]
Setting `mode=DebugMode,DebugMode.check_py=False` gives me a Python `TypeError` before starting the training. Is that a problem in the debugging codes of Theano?

```
  File "mainloop.py", line 100, in _train_epoch
    loss = self.model.train_batch(*data.values())
  File "/home/ozancag/git/Theano/theano/compile/function_module.py", line 884, in __call__
    self.fn() if output_subset is None else\
  File "/home/ozancag/git/Theano/theano/compile/debugmode.py", line 2305, in deco
    return f()
  File "/home/ozancag/git/Theano/theano/compile/debugmode.py", line 2031, in f
    new_e = e.__class__(msg)
TypeError: __init__() takes exactly 8 arguments (2 given)
```

**Comment by nouiz** [comment]
Can you give us code that reproduce the problem?

**Comment by HenryJia** [comment]
The code is Keras' MNIST CNN example:

https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py

This is my .theanorc

```
[global]
floatX = float32
device = cuda
#device = gpu

# Set allow_gc to False for faster but higher memory usage
#allow_gc = True
assert_no_cpu_op = raise
#optimizer = None
 
#[lib]
#cnmem = 512

#[nvcc]
#fastmath = True
```

When I uncomment device = cuda and comment device = gpu, the loss becomes nan, when I do the converse (old GPU backend), everything works fine

**Comment by abergeron** [comment]
I've tried with the master of libgpuarray, Theano and Keras the example you mentioned.

I get no NaNs and a result similar to the old backend:
```
$ THEANO_FLAGS=device=cuda,floatX=float32,assert_no_cpu_op=raise python mnist_cnn.py 
Using Theano backend.
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 980 (0000:05:00.0)
Downloading data from https://s3.amazonaws.com/img-datasets/mnist.pkl.gz
X_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 15s - loss: 0.3942 - acc: 0.8798 - val_loss: 0.0973 - val_acc: 0.9707
Epoch 2/12
60000/60000 [==============================] - 14s - loss: 0.1429 - acc: 0.9577 - val_loss: 0.0627 - val_acc: 0.9802
Epoch 3/12
60000/60000 [==============================] - 14s - loss: 0.1089 - acc: 0.9679 - val_loss: 0.0516 - val_acc: 0.9834
Epoch 4/12
60000/60000 [==============================] - 14s - loss: 0.0922 - acc: 0.9727 - val_loss: 0.0447 - val_acc: 0.9855
Epoch 5/12
60000/60000 [==============================] - 14s - loss: 0.0797 - acc: 0.9768 - val_loss: 0.0398 - val_acc: 0.9863
Epoch 6/12
60000/60000 [==============================] - 14s - loss: 0.0713 - acc: 0.9785 - val_loss: 0.0420 - val_acc: 0.9865
Epoch 7/12
60000/60000 [==============================] - 14s - loss: 0.0663 - acc: 0.9808 - val_loss: 0.0379 - val_acc: 0.9882
Epoch 8/12
60000/60000 [==============================] - 14s - loss: 0.0613 - acc: 0.9815 - val_loss: 0.0343 - val_acc: 0.9888
Epoch 9/12
60000/60000 [==============================] - 14s - loss: 0.0562 - acc: 0.9832 - val_loss: 0.0307 - val_acc: 0.9902
Epoch 10/12
60000/60000 [==============================] - 14s - loss: 0.0552 - acc: 0.9835 - val_loss: 0.0340 - val_acc: 0.9895
Epoch 11/12
60000/60000 [==============================] - 14s - loss: 0.0495 - acc: 0.9854 - val_loss: 0.0315 - val_acc: 0.9901
Epoch 12/12
60000/60000 [==============================] - 14s - loss: 0.0493 - acc: 0.9853 - val_loss: 0.0378 - val_acc: 0.9880
Test score: 0.0377650189095
Test accuracy: 0.988
```

For comparison I get this on the old backend:
```
THEANO_FLAGS=device=gpu,floatX=float32,assert_no_cpu_op=raise python mnist_cnn.py 
Using Theano backend.
WARNING (theano.sandbox.cuda): The cuda backend is deprecated and will be removed in the next release (v0.10).  Please switch to the gpuarray backend. You can get more information about how to switch at this URL:
 https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29

Using gpu device 0: GeForce GTX 980 (CNMeM is disabled, cuDNN 5105)
X_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 14s - loss: 0.3835 - acc: 0.8827 - val_loss: 0.0976 - val_acc: 0.9700
Epoch 2/12
60000/60000 [==============================] - 14s - loss: 0.1420 - acc: 0.9583 - val_loss: 0.0622 - val_acc: 0.9805
Epoch 3/12
60000/60000 [==============================] - 14s - loss: 0.1067 - acc: 0.9680 - val_loss: 0.0540 - val_acc: 0.9829
Epoch 4/12
60000/60000 [==============================] - 14s - loss: 0.0903 - acc: 0.9735 - val_loss: 0.0444 - val_acc: 0.9852
Epoch 5/12
60000/60000 [==============================] - 14s - loss: 0.0788 - acc: 0.9767 - val_loss: 0.0427 - val_acc: 0.9853
Epoch 6/12
60000/60000 [==============================] - 14s - loss: 0.0710 - acc: 0.9789 - val_loss: 0.0402 - val_acc: 0.9868
Epoch 7/12
60000/60000 [==============================] - 14s - loss: 0.0626 - acc: 0.9816 - val_loss: 0.0367 - val_acc: 0.9874
Epoch 8/12
60000/60000 [==============================] - 14s - loss: 0.0582 - acc: 0.9822 - val_loss: 0.0355 - val_acc: 0.9879
Epoch 9/12
60000/60000 [==============================] - 14s - loss: 0.0561 - acc: 0.9828 - val_loss: 0.0351 - val_acc: 0.9874
Epoch 10/12
60000/60000 [==============================] - 14s - loss: 0.0517 - acc: 0.9841 - val_loss: 0.0322 - val_acc: 0.9890
Epoch 11/12
60000/60000 [==============================] - 14s - loss: 0.0485 - acc: 0.9855 - val_loss: 0.0322 - val_acc: 0.9894
Epoch 12/12
60000/60000 [==============================] - 14s - loss: 0.0476 - acc: 0.9857 - val_loss: 0.0324 - val_acc: 0.9893
Test score: 0.0323955777828
Test accuracy: 0.9893
```

**Comment by nouiz** [comment]
There was a bug fix in the new back-end that was merged. Can you uninstall theano/pygpu and libgpuarray, and re-install the lastest version from github? This could fix your problem.

**Comment by HenryJia** [comment]
This is what I get
I'm running with CUDA 8, CUDNN 5.1 on Arch Linux with GTX 1080

```
[henry@henry-arch-m5a78l examples]$ THEANO_FLAGS=device=cuda,floatX=float32,assert_no_cpu_op=raise python2 mnist_cnn.py 
Using Theano backend.
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 1080 (0000:01:00.0)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.1087 - val_loss: nan - val_acc: 0.0980
Epoch 2/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 3/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 4/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 5/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 6/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 7/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 8/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 9/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 10/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 11/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 12/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Test score: nan
Test accuracy: 0.098
```

```
[henry@henry-arch-m5a78l examples]$ THEANO_FLAGS=device=gpu,floatX=float32,assert_no_cpu_op=raise python2 mnist_cnn.py 
Using Theano backend.
WARNING (theano.sandbox.cuda): The cuda backend is deprecated and will be removed in the next release (v0.10).  Please switch to the gpuarray backend. You can get more information about how to switch at this URL:
 https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29

Using gpu device 0: GeForce GTX 1080 (CNMeM is disabled, cuDNN 5105)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 6s - loss: 0.3821 - acc: 0.8825 - val_loss: 0.1015 - val_acc: 0.9679
Epoch 2/12
60000/60000 [==============================] - 6s - loss: 0.1513 - acc: 0.9553 - val_loss: 0.0672 - val_acc: 0.9786
Epoch 3/12
60000/60000 [==============================] - 6s - loss: 0.1152 - acc: 0.9657 - val_loss: 0.0543 - val_acc: 0.9817
Epoch 4/12
60000/60000 [==============================] - 7s - loss: 0.0974 - acc: 0.9715 - val_loss: 0.0475 - val_acc: 0.9842
Epoch 5/12
60000/60000 [==============================] - 6s - loss: 0.0848 - acc: 0.9756 - val_loss: 0.0446 - val_acc: 0.9859
Epoch 6/12
60000/60000 [==============================] - 6s - loss: 0.0765 - acc: 0.9768 - val_loss: 0.0425 - val_acc: 0.9856
Epoch 7/12
60000/60000 [==============================] - 7s - loss: 0.0664 - acc: 0.9800 - val_loss: 0.0367 - val_acc: 0.9873
Epoch 8/12
60000/60000 [==============================] - 6s - loss: 0.0613 - acc: 0.9817 - val_loss: 0.0360 - val_acc: 0.9885
Epoch 9/12
60000/60000 [==============================] - 6s - loss: 0.0563 - acc: 0.9833 - val_loss: 0.0338 - val_acc: 0.9884
Epoch 10/12
60000/60000 [==============================] - 7s - loss: 0.0538 - acc: 0.9840 - val_loss: 0.0314 - val_acc: 0.9887
Epoch 11/12
60000/60000 [==============================] - 6s - loss: 0.0493 - acc: 0.9849 - val_loss: 0.0323 - val_acc: 0.9886
Epoch 12/12
60000/60000 [==============================] - 6s - loss: 0.0474 - acc: 0.9861 - val_loss: 0.0305 - val_acc: 0.9893
Test score: 0.0305153276632
Test accuracy: 0.9893
```

**Comment by HenryJia** [comment]
OK will git pull and reinstall

**Comment by HenryJia** [comment]
Still no luck, same problem

**Comment by HenryJia** [comment]
@nouiz What's your setup?

**Comment by nouiz** [comment]
It is @abergeron that rerun the test. I was just suggesting something that
could explain the difference.

**Comment by Sentient07** [comment]
I think @abergeron seems to be using the Tensorflow backend using Keras. The shape of the data has the channel as the last element. In `th` backend,  it should have been (6000, 1, 28, 28)

**Comment by HenryJia** [comment]
@Sentient07 No, it does say theano backend although the format is tf which is interesting, I'll try that as well and see waht happens

**Comment by HenryJia** [comment]
@nouiz Also getting a segmentation fault at the end when python exits when I use python 3 instead of 2

```
[henry@henry-arch-m5a78l examples]$ THEANO_FLAGS=device=cuda,floatX=float32,assert_no_cpu_op=raise python mnist_cnn.py  
Using Theano backend.
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 1080 (0000:01:00.0)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.1087 - val_loss: nan - val_acc: 0.0980
Epoch 2/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 3/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 4/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 5/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 6/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 7/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 8/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 9/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 10/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 11/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 12/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Test score: nan
Test accuracy: 0.098
Segmentation fault (core dumped)
```

I'mma try shoving it into GDB and seeing what that says

**Comment by HenryJia** [comment]
Segfault seems to be coming from the GPU driver

```
[henry@henry-arch-m5a78l examples]$ gdb python
GNU gdb (GDB) 7.12.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...(no debugging symbols found)...done.
(gdb) run mnist_cnn.py 
Starting program: /usr/bin/python mnist_cnn.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7ffff3ee2700 (LWP 10617)]
[New Thread 0x7ffff16e1700 (LWP 10618)]
[New Thread 0x7ffff0ee0700 (LWP 10619)]
[New Thread 0x7fffee6df700 (LWP 10620)]
[New Thread 0x7fffebede700 (LWP 10621)]
Using Theano backend.
[Thread 0x7fffebede700 (LWP 10621) exited]
[Thread 0x7fffee6df700 (LWP 10620) exited]
[Thread 0x7ffff0ee0700 (LWP 10619) exited]
[Thread 0x7ffff16e1700 (LWP 10618) exited]
[Thread 0x7ffff3ee2700 (LWP 10617) exited]
[New Thread 0x7fffebede700 (LWP 10630)]
[New Thread 0x7fffee6df700 (LWP 10631)]
[New Thread 0x7ffff0ee0700 (LWP 10632)]
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 1080 (0000:01:00.0)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.1087 - val_loss: nan - val_acc: 0.0980
Epoch 2/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 3/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 4/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 5/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 6/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 7/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 8/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 9/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 10/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 11/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 12/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Test score: nan
Test accuracy: 0.098

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007fffd3c64805 in ?? () from /usr/lib/libcuda.so.1
(gdb) 

```

**Comment by HenryJia** [comment]
@Sentient07 @abergeron Actually I'm not sure, I just tried training with keras configured to use the TF data format with theano framework and it didn't go all NaN but it didn't train either

```
[henry@henry-arch-m5a78l examples]$ THEANO_FLAGS=device=cuda,floatX=float32,assert_no_cpu_op=raise python2 mnist_cnn.py 
Using Theano backend.
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 1080 (0000:01:00.0)
X_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 22s - loss: 2.7237 - acc: 0.1148 - val_loss: 2.1521 - val_acc: 0.0958
Epoch 2/12
60000/60000 [==============================] - 22s - loss: 2.9268 - acc: 0.0996 - val_loss: 2.1887 - val_acc: 0.0958
Epoch 3/12
60000/60000 [==============================] - 21s - loss: 2.9012 - acc: 0.0999 - val_loss: 2.1805 - val_acc: 0.0958
Epoch 4/12
60000/60000 [==============================] - 22s - loss: 2.9511 - acc: 0.1006 - val_loss: 2.1423 - val_acc: 0.0958
Epoch 5/12
60000/60000 [==============================] - 22s - loss: 2.9535 - acc: 0.1014 - val_loss: 2.1449 - val_acc: 0.0958
Epoch 6/12
60000/60000 [==============================] - 22s - loss: 2.9034 - acc: 0.1033 - val_loss: 1.9261 - val_acc: 0.0958
Epoch 7/12
60000/60000 [==============================] - 22s - loss: 2.8305 - acc: 0.1059 - val_loss: 2.0786 - val_acc: 0.0958
Epoch 8/12
60000/60000 [==============================] - 22s - loss: 2.8088 - acc: 0.1082 - val_loss: 1.9465 - val_acc: 0.0958
Epoch 9/12
60000/60000 [==============================] - 22s - loss: 2.7668 - acc: 0.1111 - val_loss: 1.9645 - val_acc: 0.0958
Epoch 10/12
60000/60000 [==============================] - 22s - loss: 2.8160 - acc: 0.1177 - val_loss: 1.9491 - val_acc: 0.0958
Epoch 11/12
60000/60000 [==============================] - 21s - loss: 2.7615 - acc: 0.1290 - val_loss: 1.9498 - val_acc: 0.0982
Epoch 12/12
60000/60000 [==============================] - 22s - loss: 2.7988 - acc: 0.1408 - val_loss: 1.9592 - val_acc: 0.1128
Test score: 1.95918997126
Test accuracy: 0.1128
```
My Keras config
```
{
    "image_dim_ordering": "tf",
    "epsilon": 1e-07,
    "floatx": "float32",
    "backend": "theano"
}

```

**Comment by Sentient07** [comment]
For the segfault, could you once clear your cache and try again ? Also, don't use the TF ordering for theano. It'd be a mess.

**Comment by HenryJia** [comment]
@Sentient07 tried theano-cache clean and then deleted everything in ~/.theano/

Still same result with TH ordering (and python2), no luck :(

**Comment by abergeron** [comment]
The segfault reminds me of an old gpuarray/Theano bug that lasted for about 6 months to make sure try to run 'bt' on the gdb prompt.

**Comment by HenryJia** [comment]
@abergeron 
The backtrace for python3

I'm not fully understanding what's causing the missing file it's reporting

```
nry@henry-arch-m5a78l examples]$ gdb python
GNU gdb (GDB) 7.12.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...(no debugging symbols found)...done.
(gdb) run mnist_cnn.py 
Starting program: /usr/bin/python mnist_cnn.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7ffff3ee2700 (LWP 6715)]
[New Thread 0x7ffff36e1700 (LWP 6716)]
[New Thread 0x7fffeeee0700 (LWP 6717)]
[New Thread 0x7fffec6df700 (LWP 6718)]
[New Thread 0x7fffe9ede700 (LWP 6719)]
Using Theano backend.
[Thread 0x7fffec6df700 (LWP 6718) exited]
[Thread 0x7fffe9ede700 (LWP 6719) exited]
[Thread 0x7fffeeee0700 (LWP 6717) exited]
[Thread 0x7ffff36e1700 (LWP 6716) exited]
[Thread 0x7ffff3ee2700 (LWP 6715) exited]
[New Thread 0x7fffe9ede700 (LWP 6794)]
[New Thread 0x7fffec6df700 (LWP 6795)]
[New Thread 0x7fffeeee0700 (LWP 6796)]
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 1080 (0000:01:00.0)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.1087 - val_loss: nan - val_acc: 0.0980
Epoch 2/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 3/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 4/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 5/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 6/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 7/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 8/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 9/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 10/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 11/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 12/12
60000/60000 [==============================] - 2s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Test score: nan
Test accuracy: 0.098

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007fffd3c64805 in ?? () from /usr/lib/libcuda.so.1
(gdb) bt
#0  0x00007fffd3c64805 in ?? () from /usr/lib/libcuda.so.1
#1  0x00007fffd3b95696 in ?? () from /usr/lib/libcuda.so.1
#2  0x00007fffd3cca992 in cuEventDestroy_v2 () from /usr/lib/libcuda.so.1
BFD: reopening /home/henry/.theano/compiledir_Linux-4.9--ARCH-x86_64-with-arch--3.6.0-64/tmpqun4ka13/m3d1cf20adb1014f04986e6a344a55bde.so: No such file or directory

BFD: reopening /home/henry/.theano/compiledir_Linux-4.9--ARCH-x86_64-with-arch--3.6.0-64/tmpqun4ka13/m3d1cf20adb1014f04986e6a344a55bde.so: No such file or directory

Can't read data for section '.eh_frame' in file '/home/henry/.theano/compiledir_Linux-4.9--ARCH-x86_64-with-arch--3.6.0-64/tmpqun4ka13/m3d1cf20adb1014f04986e6a344a55bde.so'
```

**Comment by lamblin** [comment]
@HenryJia Just to make sure, did you update your version of `libgpuarray`, and then recompile and reinstall `libgpuarray` and `pygpu`, as outlined in http://deeplearning.net/software/libgpuarray/installation.html#step-by-step-install ? I'm talking about the `libgpuarray` repository, not the module `theano.gpuarray` inside Theano.

**Comment by ozancaglayan** [comment]
Here is what I observe with that same example with the latest Theano + libgpuarray on Ubuntu 16.04, gcc/g++ v5.4.0

 - On Quadro K5200, CUDA 8-0, CuDNN v5.1 (5110, January 2017): both backends work.
 - On GTX980 with the same toolkits: new backend fails with NaNs.

```
$ THEANO_FLAGS=device=cuda,floatX=float32,assert_no_cpu_op=raise python mnist_cnn.py
Using Theano backend.
Using cuDNN version 5110 on context None
Mapped name None to device cuda: GeForce GTX 980 (0000:84:00.0)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Epoch 1/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.1086   
60000/60000 [==============================] - 3s - loss: nan - acc: 0.1087 - val_loss: nan - val_acc: 0.0980
Epoch 2/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0987
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 3/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0987
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 4/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0988
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 5/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0987
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 6/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0988
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 7/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0986
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 8/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0987
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 9/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0987
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 10/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0987
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 11/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0986
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 12/12
59776/60000 [============================>.] - ETA: 0s - loss: nan - acc: 0.0987
60000/60000 [==============================] - 3s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Test score: nan
Test accuracy: 0.098
```

@abergeron what is your CUDA version, or OS?

**Comment by HenryJia** [comment]
@ozancaglayan I'm using GCC/G++ 6.3.1 on Arch linux
@lamblin I have checked, I am using the latest libgpuarray

**Comment by abergeron** [comment]
I just ran these tests on a Debian 8 with gcc 4.9.2 and CUDA 8.0 and a GTX 980 and GTX 1080

I also changed the layout to "th" to match yours, just in case it has an impact.

With a GTX 1080
```
$ THEANO_FLAGS=device=cuda,floatX=float32,assert_no_cpu_op=raise python mnist_cnn.py
Using Theano backend.
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 1080 (0000:09:00.0)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 2s - loss: 0.3926 - acc: 0.8811 - val_loss: 0.0978 - val_acc: 0.9686
Epoch 2/12
60000/60000 [==============================] - 2s - loss: 0.1477 - acc: 0.9563 - val_loss: 0.0653 - val_acc: 0.9800
Epoch 3/12
60000/60000 [==============================] - 2s - loss: 0.1115 - acc: 0.9670 - val_loss: 0.0516 - val_acc: 0.9822
Epoch 4/12
60000/60000 [==============================] - 2s - loss: 0.0938 - acc: 0.9720 - val_loss: 0.0452 - val_acc: 0.9840
Epoch 5/12
60000/60000 [==============================] - 2s - loss: 0.0806 - acc: 0.9761 - val_loss: 0.0408 - val_acc: 0.9861
Epoch 6/12
60000/60000 [==============================] - 2s - loss: 0.0710 - acc: 0.9790 - val_loss: 0.0400 - val_acc: 0.9866
Epoch 7/12
60000/60000 [==============================] - 2s - loss: 0.0658 - acc: 0.9810 - val_loss: 0.0351 - val_acc: 0.9878
Epoch 8/12
60000/60000 [==============================] - 2s - loss: 0.0604 - acc: 0.9818 - val_loss: 0.0346 - val_acc: 0.9888
Epoch 9/12
60000/60000 [==============================] - 2s - loss: 0.0545 - acc: 0.9836 - val_loss: 0.0306 - val_acc: 0.9903
Epoch 10/12
60000/60000 [==============================] - 2s - loss: 0.0539 - acc: 0.9841 - val_loss: 0.0331 - val_acc: 0.9889
Epoch 11/12
60000/60000 [==============================] - 2s - loss: 0.0505 - acc: 0.9851 - val_loss: 0.0310 - val_acc: 0.9894
Epoch 12/12
60000/60000 [==============================] - 2s - loss: 0.0490 - acc: 0.9860 - val_loss: 0.0355 - val_acc: 0.9886
Test score: 0.03547664613
Test accuracy: 0.9886
```

And with a GTX 980 again, with the new settings:
```
$ THEANO_FLAGS=device=cuda,floatX=float32,assert_no_cpu_op=raise python mnist_cnn.py
Using Theano backend.
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX 980 (0000:05:00.0)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 3s - loss: 0.3940 - acc: 0.8804 - val_loss: 0.0997 - val_acc: 0.9686
Epoch 2/12
60000/60000 [==============================] - 3s - loss: 0.1480 - acc: 0.9562 - val_loss: 0.0651 - val_acc: 0.9801
Epoch 3/12
60000/60000 [==============================] - 3s - loss: 0.1116 - acc: 0.9671 - val_loss: 0.0519 - val_acc: 0.9832
Epoch 4/12
60000/60000 [==============================] - 3s - loss: 0.0937 - acc: 0.9721 - val_loss: 0.0456 - val_acc: 0.9842
Epoch 5/12
60000/60000 [==============================] - 3s - loss: 0.0802 - acc: 0.9760 - val_loss: 0.0412 - val_acc: 0.9858
Epoch 6/12
60000/60000 [==============================] - 3s - loss: 0.0708 - acc: 0.9787 - val_loss: 0.0413 - val_acc: 0.9866
Epoch 7/12
60000/60000 [==============================] - 3s - loss: 0.0655 - acc: 0.9810 - val_loss: 0.0353 - val_acc: 0.9878
Epoch 8/12
60000/60000 [==============================] - 3s - loss: 0.0594 - acc: 0.9820 - val_loss: 0.0349 - val_acc: 0.9886
Epoch 9/12
60000/60000 [==============================] - 3s - loss: 0.0542 - acc: 0.9835 - val_loss: 0.0310 - val_acc: 0.9899
Epoch 10/12
60000/60000 [==============================] - 3s - loss: 0.0536 - acc: 0.9841 - val_loss: 0.0330 - val_acc: 0.9897
Epoch 11/12
60000/60000 [==============================] - 3s - loss: 0.0503 - acc: 0.9847 - val_loss: 0.0311 - val_acc: 0.9894
Epoch 12/12
60000/60000 [==============================] - 3s - loss: 0.0487 - acc: 0.9860 - val_loss: 0.0363 - val_acc: 0.9880
Test score: 0.0363492626627
Test accuracy: 0.988
```

I really can't manage to reproduce this.  Just to make sure I am at commit https://github.com/fchollet/keras/commit/14f35ab05551ec0b4531489cd4c48a2dd322020e, https://github.com/Theano/Theano/commit/5bb6abe5f84f0880a770d4c94c94c578d6153e34 and https://github.com/Theano/libgpuarray/commit/8cc349db7f66808625aaa59196dadd8a77a718a6.

**Comment by HenryJia** [comment]
@abergeron Have you tried using newer versions of GCC, I'm on 6.3.1

**Comment by abergeron** [comment]
Can you make a singularity[1] image with your setup?  This would be helpful to try and reproduce the problem with your exact environment.

1: http://singularity.lbl.gov/

**Comment by abergeron** [comment]
I tried with gcc 6.2.0 (which is the highest version available as a ppa) on a GTX 750 (I can't try it on the 980 or 1080 since they are lab machines where we can't install arbitrary stuff) and there are still no NaNs.  With the singularity image I could try on the 1080 or 980.

Barring compiler bugs, I would be surprised if the version of gcc has a significant impact on the GPU results.

**Comment by HenryJia** [comment]
Just tried it on a Ubuntu 16.04 server (AWS EC2 P2 instance with 1 x K80).

GCC version 5.4.0-6
CMake 3.5.1
NumPy 1.12.0
SciPy 0.18.1
Cython 0.25.2
Mako 1.0.6
Nose 1.3.7
MarkupSafe 0.23

No CuDNN

Did not get any NaNs.

I'll compare my setups to see what's going wrong

**Comment by ozancaglayan** [comment]
I tried with no CuDNN as well on Ubuntu 16.04 but still NaNs. I also compared the graphs of the backends, they are same. Finally, let me remind you that in my case `optimizer=None` fixes the issue.

**Comment by ozancaglayan** [comment]
@HenryJia well the K80 is a Kepler card but the bug happens on Maxwell series I think - at least for me and probably for you as well -

**Comment by abergeron** [comment]
If you use optimizer=None, you will not get any NaN, but the code will run on the CPU.  You can try optimizer=fast_compile to enable a minimum of optimization that mostly move stuff to the GPU.

Also for the source of the problem, I just had a flash.  Are you using the distribution-provided drivers or the ones from Nvidia?

**Comment by ozancaglayan** [comment]
I am using 375.26 from Nvidia's website. Nope, `fast_compile` is broken as well.
I also tried with/without anaconda, and python 2.x/3.x. All broken.

**Comment by HenryJia** [comment]
@ozancaglayan Hmmm, just noticed, the Kepler K80 is running driver 367.57, I'll see what happens when I downgrade the driver

**Comment by HenryJia** [comment]
Scratch that, that needs downgrading the linux kernel

**Comment by abergeron** [comment]
I'm on 365.57.  I can try with the newer driver, but if that breaks I would consider it an Nvidia bug :)

**Comment by abergeron** [comment]
I mean 367.57

**Comment by HenryJia** [comment]
@abergeron Thanks, I'm also on 375.26 BTW

**Comment by abergeron** [comment]
I just tried with 375.26 and I get NaNs with my GTX 750.  I'll look into it.

**Comment by HenryJia** [comment]
Thanks :+1:

**Comment by abergeron** [comment]
I've got someone else report to me that 373.06 works.  Maybe you can install that without downgrading your kernel.  Otherwise you can use the runfile to install the driver.

In any case the 375 series is definitely broken.  It fails doing a simple sum.  We'll make a bug report to nvidia about this.

**Comment by ozancaglayan** [comment]
FYI, I gave the new beta driver 378.09 a try, still does not work.

**Comment by HenryJia** [comment]
Switched back to 370.28 driver instead, no Arch Linux package for 373.06, ended up downgrading kernel and Xorg in the end anyway.

Everything works as expected, although python3 still segfaults on exit() (less of an issue).

**Comment by nouiz** [comment]
We merged the detection of the bad driver in libgpuarray, so I close this issue. For the seg fault, can you open a new issue with information on how to reproduce it?

thanks

**Comment by alyato** [comment]
hi @abergeron  i upgrade the theano to `0.9 dev`,but it show me 

> ValueError: You are tring to use the old GPU back-end. It was removed from Theano. Use device=cuda* now. 

how do you setup your theano?
I use the keras base on theano. 
The os is ubuntu 14.04 lts.

**Comment by abergeron** [comment]
Do what the message says and switch devices.  We removed the old backed from the development version.

**Comment by nouiz** [comment]
Here is the doc to do the full convert in case just changing the flag isn't
enough:

https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29
